Make library flag name resolution concurrent in posix linking

domgetter opened the pull request:
This makes library name resolution concurrent within the posix linking step.

This reduces the time taken in the linking step of compilation by 30-70% depending on the size of the project

See #11888 for more timing information.

The code does not take care to maintain the order of the flags, so if the order matters, let me know and I can add indices and sort them after the resolution.

Sija commented on src/compiler/crystal/codegen/link.cr:
```diff
@@ -145,23 +146,32 @@ module Crystal
-        elsif (lib_name = ann.lib)
-          flags << Process.quote_posix("-l#{lib_name}")
+        spawn do
+          next_flag = ""
+          if ldflags = ann.ldflags
+            next_flag = ldflags
+          end
+
+          # First, check pkg-config for the pkg-config module name if provided, then
+          # check pkg-config with the lib name, then fall back to -lname
+          if (pkg_config_name = ann.pkg_config) && (flag = pkg_config(pkg_config_name, static_build))
+            next_flag = flag
+          elsif (lib_name = ann.lib) && (flag = pkg_config(lib_name, static_build))
+            next_flag = flag
+          elsif (lib_name = ann.lib)
+            next_flag = Process.quote_posix("-l#{lib_name}")
+          end
+
+          if framework = ann.framework
+            next_flag = "-framework" + Process.quote_posix(framework)
+          end
+
+          flags_channel.send next_flag
```
I see 2 problems with this impl:

- it would send an empty flag if none of the conditions match
- it overwrites the previously assigned flag (`next_flag`), in previous impl. it always added the matched flag via `flags <<`

Sija commented:
I see 2 problems with this impl:

- it would send an empty flag if none of the conditions match
- it overwrites the previously assigned flag (`next_flag`), in previous impl. it always added the matched flag via `flags <<`

domgetter commented:
Okay so I'll need to send an `Array(String)` over the channel, and make a temporary list of `next_flags` within the thread, and then concat/append the list as they are received.

Sija commented on src/compiler/crystal/codegen/link.cr:
```diff
@@ -145,23 +146,32 @@ module Crystal
-          flags << "-framework" << Process.quote_posix(framework)
-        end
+      link_annotations.size.times do
+        flags += flags_channel.receive
```
Use `Array#concat` to avoid intermediate allocation.

```suggestion
        flags.concat(flags_channel.receive)
```

straight-shoota commented:
> The code does not take care to maintain the order of the flags, so if the order matters, let me know and I can add indices and sort them after the resolution.

Yes, the order of flags matters. See https://github.com/crystal-lang/crystal/blob/396324414bd57da362f954cd5f585bdbb33d92b1/src/gc/boehm.cr#L16-L17 for example.

straight-shoota commented:
I believe the implementation can be simplified a bit. The values of `ann.ldflags` and `ann.framework` are fixed and don't need to be resolved concurrently. That leaves only a single value per annotation to be resolved via `pkg-config`, thus `next_flags` array is unnecessary.

Maybe it would be best to separate the entire logic into two independent stages: First, resolve package names via `pkg-config`. Second, collect the `flags` arguments. As a simple solution, the resolved package names can be stored in the `Annotation` instance. We need to add a property for this. 

Simplified code example:
```cr
channel = Channel(Nil).new
link_annotations.each do |ann|
  spawn do
    ann.resolved_name = resolve_pkg_config(ann)
    channel.send(nil)
  end
end
link_annotations.each do
  channel.receive
end
link_annotations.reverse_each do
  flags << ann.resolved_name
end
```
I ommitted some details (such as handling of `ldflags` and `framework`) for brevity.

domgetter commented:
> Yes, the order of flags matters. For example:

Got it, I've updated to maintain the order.

domgetter commented:
> Maybe it would be best to separate the entire logic into two independent stages: First, resolve package names via `pkg-config`. Second, collect the `flags` arguments. As a simple solution, the resolved package names can be stored in the `Annotation` instance. We need to add a property for this.

Do you think this is something that should make it into this pull request, or should we save that for a future one?

straight-shoota commented:
If you're up to it, I'd prefer that. You might wait a moment for other comments first, though.

asterite commented:
I think it should be done in a separate PR. This PR already improves things. I don't see why we need to add more to it right now, making its review harder.

asterite commented:
Remember people: a small step forward is better than staying still.

asterite commented:
And in this case this is a big step forward

straight-shoota commented:
@asterite It's just that the solution I have in mind is probably better and would be similar to implement, but with a smaller diff overall. So there's less to review (both for this change and the next). I'd rather go that extra step directly instead of having to deal with a perliminary.
I applied my changes on top of these to see the difference: https://github.com/straight-shoota/crystal/commit/7a6127883f50ee0b03417c6a205940ee30370bc6

asterite commented:
> I applied my changes on top of these to see the difference: https://github.com/straight-shoota/crystal/commit/7a6127883f50ee0b03417c6a205940ee30370bc6

Great! And that actually helps understand how I'd like things to be. I'll explain by giving an example of my past contributions to Ruby.

- I found a couple of bugs in Ruby
- I reported the bug
- I send a PR to fix the bug
- The PR was merged right away
- After that, another commit came in that improved the formatting and adjusted things to the way the core team wanted things to be

This was a really nice experience! I was able to contribute, giving an initial idea/patch, and the rest was taken care by someone else. No needless discussions. No burden put on me asking me to change this or that.

This is what I'd also like to happen in Crystal. Here's a PR that greatly improves the compile times. Do we take it? No! We ask for a better solution. Will the PR author do that? Maybe! But why put the burden on them? They already contributed in a great way. And asking things with words and having the PR author translate that request to the actual code that was asked is really hard! It's so much easier to just accept this PR and send a follow up PR with the exact code that we want. Heck, you even did that, but in a separate commit! Now the PR author has to incorporate that changes in this PR, but if we did things like Ruby that would already be in master.

This is just my opinion! Without a face or emojis I might look like angry or requesting forcefully, but I'm just giving my opinion :-)

What do you think?

I think I mentioned this in the past several times, so feel free to ignore me here and maybe I can open another issue or forum topic discussing this.

domgetter commented:
I'm perfectly happy just to have caused the eventual improvement.

@straight-shoota's simplification looks awesome and it teases apart the purpose of the code, instead of just heavy-handedly making it faster like mine did.

Forgive for not being as experienced with contributions, but I'm not really sure how to incorporate the new code into this PR; I don't know if there's some procedure or something. With @Sija's there was a button to merge it in.

If everyone likes it, I'd move for @straight-shoota's code to make it in, and we can move forward to find other improvements, like resolving the lib names earlier in the pipeline.

oprypin commented:
(Note that I have not reviewed any implementation, just sharing a tip)

@domgetter  How to merge this in:
Considering that this commit was shared: 7a6127883f50ee0b03417c6a205940ee30370bc6

```bash
git fetch origin 7a6127883f50ee0b03417c6a205940ee30370bc6
git merge 7a6127883f50ee0b03417c6a205940ee30370bc6
git push
```

The trick that commits are shared across all repos is GitHub-specific; if it doesn't work, this line would be needed instead:

```bash
git fetch https://github.com/straight-shoota/crystal 7a6127883f50ee0b03417c6a205940ee30370bc6
```

domgetter commented:
I have incorporated @straight-shoota's implementation and also implemented his idea of kicking off the name resolution earlier in the compile pipeline, which saves a little bit more time.

You can see some more rigorous performance analysis in my comment here: https://github.com/crystal-lang/crystal/issues/11888#issuecomment-1069346391

Long story short: most of the hang-up was due to being on WSL1 (which will be a big jump for anyone still using that), but even on WSL2, there is measurably ~14% time reduction in the linker step with all changes added in (about 60ms).

I'll be using WSL2 from now on for Crystal dev, that's for sure.

Also, I recommend installing and using `mold`, which can be used easily with `CC="clang -fuse-ld=mold" crystal build ...`, as this cuts down linker time by another good chunk (linking time when compiling the compiler takes only 0.4s even on WSL2, instead of 3-4s).

straight-shoota commented:
I don't like pulling the entire `linker_command` call up in the concurrent execution flow. This could easily lead to weird concurrency issues if we do stuff in that method that might interfere with other stages of the linking process. I don't think that's the case currently, but I'm not sure. And it could become a problem in the future. I'd rather keep the concurrently running tasks as small as possible. That is, only the direct link annotation resolving for each `LinkAnnotation` instance.

domgetter commented:
> I don't like pulling the entire `linker_command` call up in the concurrent execution flow. This could easily lead to weird concurrency issues if we do stuff in that method that might interfere with other stages of the linking process.

The call to `linker_command_channel.receive` happens before anything else in the linking process, so it won't be concurrent with anything else there.

nobodywasishere commented:
I think this PR could be really useful, can we bring it up to speed and review / merge it?
